Cli/ associate words in quotes into a single item (token)

mrestuccia1 opened the pull request:
Currently, there is a limit of four tokens per input, which limits the possible inputs. While there is no need for it right now, we want to be able to say, for example, go to "ring of fire", and have that be counted as three tokens instead of five. This will allow for more tokens, but also for games to have special names beyond north, south etc... We edited the current parsing framework such that if the command line input contains something in quotes, whatever is inside is considered its own token, no matter how many words are inside.

milesgardner committed:
Adding first parser.c changes

milesgardner committed:
More changes to parser.c

Commit:
attempt

milesgardner committed:
small changes

milesgardner committed:
Merge branch 'Parse-Tokens' of github.com:uchicago-cs/chiventure into Parse-Tokens

Commit:
new code

milesgardner committed:
more additions for quotation parsing

Commit:
added quotes functionality

milesgardner committed:
Adding additional files

milesgardner committed:
Resolving merge conflicts

Commit:
added functionality for quotes to be at the beginning

Commit:
added quotes at the beginning of input handling

majorsylvie commented on src/cli/src/parser.c:
```diff
@@ -105,27 +106,80 @@ char **parse(char *input)
     }
 
-    char *token = strtok(input, " ");
+    char *token2 = strtok(input, "\"");
```
tokenizing on quotation mark

majorsylvie commented on src/cli/src/parser.c:
```diff
@@ -105,27 +106,80 @@ char **parse(char *input)
+    int pos = 0;
 
-    for(int i = 0; i < TOKEN_LIST_SIZE; i++)
+    for(int i = 0; i < 4; i++)
```
why change this from being the constant to 4? Should this not be linked?

mrestuccia1 replied:
This is correct, we switched it initially to test the function inside a sandbox, and never switched it back. This change has been made, and all the for loops now have dynamic variables.

majorsylvie commented on src/cli/src/parser.c:
```diff
@@ -105,27 +106,80 @@ char **parse(char *input)
-    //If there are more than 4 words, parser returns NULL and does not attempt
-    //to pass the first four words as tokens
-    if(token != NULL)
+    if (input[0] != '\"')
```
should add some kind of description here about what the loop is doing.

mrestuccia1 replied:
We have added descriptive comments for everything we added, so this change should be fixed now.

majorsylvie commented on src/cli/src/parser.c:
```diff
@@ -105,27 +106,80 @@ char **parse(char *input)
     {
+        char *token = strtok(input, " ");
+
+        for(int i = 0; i < 4; i++)
```
Again, why 4? It is not clear to me. 
If this is meant to be the word limit constant, then keep it as such

jacknugent1529 replied:
@fkbad Is correct. Avoid using "[magic numbers](https://uchicago-cs.github.io/dev-guide/style_guide_c.html#avoid-magic-numbers)" because it is often unclear what the numbers mean (especially if someone changes them in the future). Instead use descriptive constants

mrestuccia1 replied:
This change has been made, there are descriptions for all our new added code, and we changed the "magic numbers" back to the dynamic variables they had been set to in the previous version.

majorsylvie reviewed (changes requested):
Please add some description within your code (particularly at the commented portions) so that it is more clear what you code is meant to be doing.

I already know it is supposed to parse for text within a quote, but it is not obvious when you are looking.

Commit:
added changes suggested in Faruk's PR

mrestuccia1 commented:
As suggested by Faruk in his review, we added comments and changed the number 4 back to the pre-defined variable TOKEN_LIST_SIZE. The code still passes all the tests and added the functionality described above. We will check with Faruk one more time, but the next step will be to get the code looked over by the TA

jacknugent1529 commented on src/cli/src/parser.c:
```diff
@@ -105,27 +106,93 @@ char **parse(char *input)
     words = (char**)malloc(sizeof(char*)*TOKEN_LIST_SIZE);
+    char **temp;
+    temp = (char**)malloc(sizeof(char*)*TOKEN_LIST_SIZE);
+    char **temp2;
```
More descriptive names (rather than `temp` and `temp2`) would make this code much easier to read. Also, if names are not enough, give a brief comment explaining why there needs to be 3 of these arrays.

mrestuccia1 replied:
We reverted the code to only need two arrays, one named words, which is the array of tokens tokenized by spaces, and one named words_quotes, which is the array of tokens tokenized by quotes.

jacknugent1529 commented on src/cli/src/parser.c:
```diff
@@ -105,27 +106,93 @@ char **parse(char *input)
+
+        if(token3 != NULL)
+        {
+        return NULL;
```
style: indentation, and include a space after `if`

mrestuccia1 replied:
The indentation has been added.

jacknugent1529 replied:
I mean that the `return NULL` line should be indented once more based on its location in the curly braces. And the space between `if` and `(` doesn't seem to have been added.

mrestuccia1 replied:
The indentation should now be added for all instances of return NULL, and spaces between all instances of if and "(" have been added.

jacknugent1529 reviewed (commented):
The main thing to change is the descriptiveness of the names. Everything else looks good, although I'll do a more thorough review once this is done. Also, since this is an integral part of chiventure, testing is especially important. So, how have you tested this code and can we be confident that it interpreting the quotes in this manner won't break something elsewhere?

Commit:
committing before switching branch

milesgardner commented:
General functionality has been completed.  Unit tests have been implemented, and we believe that tokens surrounded by quotations is ready to be merged into the main branch.

mrestuccia1 commented:
> Please add some description within your code (particularly at the commented portions) so that it is more clear what you code is meant to be doing.
> 
> I already know it is supposed to parse for text within a quote, but it is not obvious when you are looking.

jacknugent1529 commented on src/cli/src/parser.c:
```diff
@@ -105,27 +106,98 @@ char **parse(char *input)
     }
 
-    char *token = strtok(input, " ");
+    char *token2 = strtok(input, "\"");
```
I still think this naming scheme is a bit confusing. What does "token2" mean in this context?

mrestuccia1 replied:
Just changed token2 to token_quotes, since it is the token separated by quotes.

jacknugent1529 commented on src/cli/src/parser.c:
```diff
@@ -105,27 +106,98 @@ char **parse(char *input)
+        char *token = strtok(input, " ");
+
+        //Populates words array with tokens separated by space (" ") character
+        for(int i = 0; i < 4; i++)
```
style: space between `for` and `(` (`for (...`). Also, replace 4 with a constant

mrestuccia1 replied:
Added the spaced between for and ( in all instances, and changed 4 to the dynamic variable it was previously set to.

milesgardner committed:
Adding unit tests for quote tokens

milesgardner committed:
Merge branch 'Parse-Tokens' of github.com:uchicago-cs/chiventure into Parse-Tokens

milesgardner committed:
Adding unit tests for parsing quotes

milesgardner commented:
Added unit tests as requested through slack.

jacknugent1529 commented on src/cli/src/parser.c:
```diff
@@ -105,27 +106,98 @@ char **parse(char *input)
+        return words;
+
+    //If the first character of the input is "
+    }else{
```
style: spaces around `else`

mrestuccia1 replied:
Added the spaces around else, as suggested.

jacknugent1529 reviewed (commented):
It looks mostly good and the tests are a nice addition. There's just a few style issues remaining

mrestuccia1 commented:
All requested edits have been made, we believe the pull request should now be good to go.

jacknugent1529 committed:
Merge branch 'dev' into Parse-Tokens

Commit:
added requested changes

Commit:
Merge branch 'Parse-Tokens' of github.com:uchicago-cs/chiventure into Parse-Tokens

jacknugent1529 reviewed (commented):
Nice job making the changes! It looks good to me

jacknugent1529 commented:
PR Score: Excellent

Comments: Excellent work on this PR! The description is good, and you addressed each of the suggestions well. This is not a rubric item, but the commit "discipline" could be improved to give insightful yet concise commit messages and commits in places that logically make sense. 

<!-- {"score": "excellent", "minor": false} -->
